iroh/src/commands/provide.rs

```diff
@@ -164,15 +165,30 @@ async fn get_keypair(key: Option<PathBuf>) -> Result<Keypair> {
         Some(key_path) => {
             if key_path.exists() {
                 let keystr = tokio::fs::read(key_path).await?;
-                let keypair = Keypair::try_from_openssh(keystr)?;
+                let keypair = Keypair::try_from_openssh(keystr).context("invalid keyfile")?;
                 Ok(keypair)
             } else {
                 let keypair = Keypair::generate();
                 let ser_key = keypair.to_openssh()?;
                 if let Some(parent) = key_path.parent() {
                     tokio::fs::create_dir_all(parent).await?;
                 }
-                tokio::fs::write(key_path, ser_key).await?;
+                // write to tempfile
+                let (file, temp_file_path) = tempfile::NamedTempFile::new()
```

Comment:
I think you need `NamedTempFile::new_in()` to put it in the same directory as the destination.  Otherwise the tempfile could be on a different filesystem and copying is still not atomic.

Reply:
fixed